### ISSUE TITLE
reproduce opensource data using qlib implementation

### ISSUE THREAD

**Issue by zzj2404** [SJTU-DMTai/MASTER#36]
these are my steps:
1. Downloaded dataset from: https://github.com/chenditc/investment_data/releases/tag/2025-05-28
2. Cloned the Qlib implementation from [qlib implementation](https://github.com/SJTU-Quant/qlib/tree/main/examples/benchmarks/MASTER) and install env using config.sh
3. Modified **workflow_config_master_Alpha158.yaml** , make timespan and learning rate same to paper
4. Modified **pytorch_master_ts.py**, add drop_extreme(label) in train_epoch，and zscore(label) in test_epoch

![Image](https://github.com/user-attachments/assets/62cf0911-0624-46a4-bbda-563097f3c703)

![Image](https://github.com/user-attachments/assets/bd708505-fb34-4061-acde-f7ef2c09fbdd)

5. Modified **get_feature_config** following readme
6. Modified **pytorch_master_ts.py**, copy MASTER self.layers in this repo to replace qlib implementation's layer and load csi300_opensource_0.pkl in **main.py**, Modified **pytorch_master_ts.predict** param use_pretrained to False

then i get a good result on seed 0:

![Image](https://github.com/user-attachments/assets/287339eb-f3cd-440d-b106-a59f51b7dc8a)

but when i use step1~5 to train and test , i get a bad result on seed 0:

![Image](https://github.com/user-attachments/assets/549384c1-aca2-493e-ad44-e8eadd7bb222)

Since using the author's parameters on the same dataset can reproduce the results, the issue might be with my training setup? However, I haven't been able to identify the problem. I would be extremely grateful if someone could provide an answer!

**Comment by LITONG99** [comment]
Hi, 

It seems you are manually implementing `CSZScoreNorm` in the code. Since you did not mention disabling the `CSRankNorm` in the .yaml, could you please check the `workflow_config_master_Alpha158.yaml` to make sure you removed `CSRankNorm`? You may find this [closed issue](https://github.com/SJTU-DMTai/MASTER/issues/8) helpful.

Please let me know if your problem is solved.

**Comment by LITONG99** [comment]
The .yaml seems correct to me.

Could you please check lines 280-283 in [pytorch_master_ts.py](https://github.com/SJTU-DMTai/qlib/blob/main/qlib/contrib/model/pytorch_master_ts.py)? The $\beta$ of the gate is supposed to be the same as in our paper, which is
> if self.market == 'csi300': self.beta = 5
else: self.beta = 2

Since I am not responsible for the maintenance of the Qlib version, I also suggest double-checking other hard-coded hyperparameters. I will try to check it by myself in a few days and update one copy of `pytorch_master_ts.py` in this repo later.

**Comment by zzj2404** [comment]
i think this result make sense?

![Image](https://github.com/user-attachments/assets/f60c0880-26e3-4307-8e00-a99a59ebc439)

![Image](https://github.com/user-attachments/assets/95fac4e9-e526-4608-87aa-2141f93abafd)

Apart from the difference in self.beta compared to the paper, the logic for saving parameters in the fit function is also different. The qlib version saves parameters based on validation loss, while this repository saves them based on training loss. After modifying these two aspects, I obtained the results shown above, although there is still a gap in AR and IR. 

Finally, thank you for your patient and thorough explanations!

**Comment by LITONG99** [comment]
It is glad to hear!
The obtained AR and IR seem reasonable to me, since it is within the the experimental variance range. I made some update under folder `/qlib-update` and `Reademe.md`, which may be helpful in further eliminate the gap.

**Comment by LPeng625** [comment]
> 以下是我的步骤：
> 
> 1. 下载数据集来源：https://github.com/chenditc/investment_data/releases/tag/2025-05-28
> 2. 用 [qlib 实现](https://github.com/SJTU-Quant/qlib/tree/main/examples/benchmarks/MASTER)克隆了 qlib 实现，并用 config.sh 安装环境
> 3. 修改**workflow_config_master_Alpha158.yaml**，时间跨度和学习速度都与纸张保持一致
> 4. 修改**后的pytorch_master_ts.py，train_epoch**中添加drop_extreme（label），在test_epoch中添加zscore（label）
> 
> ![Image](https://github.com/user-attachments/assets/62cf0911-0624-46a4-bbda-563097f3c703)
> 
> ![Image](https://github.com/user-attachments/assets/bd708505-fb34-4061-acde-f7ef2c09fbdd)
> 
> 5. 修改**get_feature_config**后续说明
> 6. 修改**后pytorch_master_ts.py**，复制本仓库中的 MASTER self.layers 以替换 qlib 实现的层，并在 **main.py** 加载 csi300_opensource_0.pkl，修改后的 **pytorch_master_ts.predict** param 为 use_pretrained False
> 
> 然后我对种子0有个不错的结果：
> 
> ![Image](https://github.com/user-attachments/assets/287339eb-f3cd-440d-b106-a59f51b7dc8a)
> 
> 但当我用step1~5进行训练和测试时，种子0却出现了不好的结果：
> 
> ![Image](https://github.com/user-attachments/assets/549384c1-aca2-493e-ad44-e8eadd7bb222)
> 
> 既然在同一数据集上使用作者的参数可以复现结果，问题可能出在我的训练设置上？不过，我一直没能找到问题所在。如果有人能给我答案，我将非常感激！

您好，我也遇到了ir低于1的情况，可以分享下您的配置文件吗？不胜感激。

**Comment by LPeng625** [comment]
> i think this result make sense?
> 
> ![Image](https://github.com/user-attachments/assets/f60c0880-26e3-4307-8e00-a99a59ebc439)
> 
> ![Image](https://github.com/user-attachments/assets/95fac4e9-e526-4608-87aa-2141f93abafd)
> 
> Apart from the difference in self.beta compared to the paper, the logic for saving parameters in the fit function is also different. The qlib version saves parameters based on validation loss, while this repository saves them based on training loss. After modifying these two aspects, I obtained the results shown above, although there is still a gap in AR and IR.
> 

<img width="914" height="135" alt="Image" src="https://github.com/user-attachments/assets/623617b8-8094-4d78-87ec-1176aff6fb18" />

Compared with the results provided by the author, the first four indicators are almost the same, and the last two indicators are significantly different. Please tell me how to deal with it, thank you and look forward to your reply.
